File: service/game/actions/work.py
import time

from service.game.actions.base import Command
from service.game.actions.contracts import TradeContract


DEVELOPMENT_OUTPUT = {"Farm": "food", "Woods": "wood", "Mine": "iron"}


class CommitWorkCommand(Command):
    def execute(self, game_state, player):
        if player.health in ["sick", "recovering", "dead"]:
            return False

        job_data = self.payload.get("job")
        if not job_data:
            return False

        action_id = job_data.get("action_id")
        dev_id = job_data.get("development", {}).get("id")
        live_dev = game_state.developments.get(dev_id)
        if live_dev and getattr(live_dev, "is_contested", False):
            return False

        if action_id:
            chosen_action = game_state.contract_factory.find_contract(
                action_id
            )
            if chosen_action:
                chosen_action.status = "COMPLETED"
                if chosen_action.type == "EMPLOYMENT":
                    wage_trade = TradeContract(
                        initiator_id=chosen_action.target_id,
                        target_id=chosen_action.initiator_id,
                        offer_items={
                            chosen_action.wage_type: chosen_action.wage
                        },
                        request_items={},
                    )
                    wage_trade.status = "ACCEPTED"
                    wage_trade.target_finalized = True
                    game_state.contract_factory._add_contract_to_players(
                        wage_trade
                    )

                for action in list(player.actions.values()):
                    if (
                        getattr(action, "type", None) == "EMPLOYMENT"
                        and action.status == "ACCEPTED"
                        and action.id != action_id
                    ):
                        action.status = "CANCELED"

        player.committed_action = job_data
        return True


def resolve_work_phase(game_state):
    for player in game_state.players.values():
        if player.health in ["sick", "recovering"]:
            continue
        committed_action = getattr(player, "committed_action", None)
        if not committed_action or not isinstance(committed_action, dict):
            continue

        development = committed_action.get("development", {})
        owner_id = development.get("owner_id")
        development_type = development.get("type")
        development_level = int(development.get("level", 1))
        owner = game_state.players.get(owner_id)
        resource = DEVELOPMENT_OUTPUT.get(development_type)
        if not owner or not resource:
            continue

        owner.resources[resource] = (
            owner.resources.get(resource, 0) + development_level
        )
        if owner_id != player.session_id:
            owner.add_timeline_event(
                "LABOR_EXPLOITED",
                {
                    "worker": player.session_id,
                    "yield": development_level,
                    "type": resource,
                },
            )


def start_work_phase(game_state):
    if game_state.status != "RUNNING":
        time.sleep(3)
        return

    for player in game_state.players.values():
        player.available_work = []

    for development in game_state.developments.values():
        if getattr(development, "is_contested", False):
            continue
        owner = game_state.players.get(development.owner)
        if not owner:
            continue

        resource = DEVELOPMENT_OUTPUT.get(development.type, "food")
        owner.available_work.append(
            {
                "development": (
                    development.to_dict()
                    if hasattr(development, "to_dict")
                    else development.__dict__
                ),
                "wage": development.level,
                "wage_type": resource,
                "employer_id": owner.session_id,
                "action_id": None,
            }
        )

File: service/src/game/actions/work.ts
import {
  commitWorkPayloadSchema,
  type EmploymentActionDTO,
  type GameStateDTO,
  type PlayerDTO,
  type WorkActionDTO,
} from "@takes-a-village/shared";

interface WorkPlayer {
  readonly id: string;
  readonly actions: PlayerDTO["actions"];
  readonly resources: PlayerDTO["resources"];
  readonly timeline: PlayerDTO["timeline"];
  health: PlayerDTO["health"];
  finishedPhase: boolean;
  committedAction: PlayerDTO["committed_action"];
}

interface WorkGame {
  readonly players: Map<string, WorkPlayer>;
  readonly developments: Map<string, GameStateDTO["developments"][number]>;
  readonly phase: GameStateDTO["phase"];
}

export function handleWorkAction(
  game: WorkGame,
  player: WorkPlayer,
  command: unknown,
  payload: Record<string, unknown>,
): boolean {
  if (command !== "COMMIT_WORK" || game.phase !== "WORK" || player.health !== "healthy") return false;
  const parsed = commitWorkPayloadSchema.safeParse(payload);
  if (!parsed.success) return false;
  const job = parsed.data.job;
  const liveDevelopment = game.developments.get(job.development.id);
  if (liveDevelopment?.is_contested) return false;
  if (job.action_id) {
    const employment = player.actions.find(
      (action): action is EmploymentActionDTO => action.id === job.action_id && action.type === "EMPLOYMENT",
    );
    if (!employment || employment.status !== "ACCEPTED") return false;
    employment.status = "COMPLETED";
  }
  player.committedAction = job;
  player.finishedPhase = true;
  return true;
}

export function resolveWork(game: WorkGame): void {
  for (const worker of game.players.values()) {
    if (worker.health === "sick" || worker.health === "recovering" || worker.health === "dead") continue;
    const job = asWorkAction(worker.committedAction);
    if (!job) continue;
    const owner = game.players.get(job.development.owner_id);
    const resource = developmentOutput(job.development.type);
    if (!owner || !resource) continue;
    owner.resources[resource] += job.development.level;
    if (owner.id !== worker.id) {
      owner.timeline.push({
        type: "LABOR_EXPLOITED",
        data: { worker: worker.id, yield: job.development.level, type: resource },
      });
    }
  }
  for (const player of game.players.values()) {
    player.committedAction = null;
    for (let index = player.actions.length - 1; index >= 0; index -= 1) {
      if (player.actions[index]?.type === "EMPLOYMENT") player.actions.splice(index, 1);
    }
  }
}

function asWorkAction(value: PlayerDTO["committed_action"]): WorkActionDTO | null {
  if (!value || !("development" in value)) return null;
  return value;
}

function developmentOutput(type: GameStateDTO["developments"][number]["type"]): keyof PlayerDTO["resources"] {
  if (type === "Farm") return "food";
  if (type === "Woods") return "wood";
  return "iron";
}
